### PR TITLE
change readCache to final

### PR DESCRIPTION
**Pull request by no2chem**
This PR changes the readCache from being static (global) to a final instance
field, preventing possible odd behaviors from occurring and runtimes sharing
caches (which they shouldn't, as they should operate independently).

In addition, this patch now also contains a fix to the checkpoint entry type, and
since we have sort of a chicken-and-egg problem of unit tests and fixes
 (the unit tests are "fixed" by this PR, but the code is "fixed" by fixes to the 
checkpointing entries), they are all in this PR.

### CODE REVIEW

**Comment by no2chem** [comment]
@slfritchie the checkpoint smoke tests seem to fail this test on some nettybytebuf issue.. any idea why this small change might cause this?

**Comment by no2chem** [comment]
Ok, @slfritchie the issue seemed to be that checkpointing entries used the SMR type, when they should have had their own type defined. The checkpointing mechanism won't work in a real setup without that.

Could you review those changes to see if they look okay to you?

File: runtime/src/main/java/org/corfudb/protocols/logprotocol/CheckpointEntry.java [hunk 1]
```diff
@@ -99,7 +99,7 @@ public byte asByte() {
 
     public CheckpointEntry(CheckpointEntryType type, String authorID, UUID checkpointID,
                            Map<CheckpointDictKey,String> dict, MultiSMREntry smrEntries) {
-        super(LogEntryType.SMR);
+        super(LogEntryType.CHECKPOINT);
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/protocols/logprotocol/CheckpointEntry.java]: Why do you need this change?

File: runtime/src/main/java/org/corfudb/protocols/logprotocol/LogEntry.java [hunk 2]
```diff
@@ -113,7 +113,8 @@ public boolean isMutation(UUID stream) {
         SMR(1, SMREntry.class),
         STREAM_COW(4, StreamCOWEntry.class),
         MULTIOBJSMR(7, MultiObjectSMREntry.class),
-        MULTISMR(8, MultiSMREntry.class);
+        MULTISMR(8, MultiSMREntry.class),
+        CHECKPOINT(10, CheckpointEntry.class);
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/protocols/logprotocol/LogEntry.java]: Why is this related to making the caches non-static ?

File: runtime/src/main/java/org/corfudb/protocols/logprotocol/CheckpointEntry.java (hunk 1 above)
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/protocols/logprotocol/CheckpointEntry.java]: follow the PR dicussion

File: runtime/src/main/java/org/corfudb/protocols/logprotocol/LogEntry.java (hunk 2 above)
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/protocols/logprotocol/LogEntry.java]: also follow the PR discussion, due to a failing test

**Comment by no2chem** [comment]
@slfritchie yes

Forcing per-instance read caches seems to give us more test coverage as it prevents our single-JVM unit tests from falsely sharing caches. Once we have assurance that performance is not affected, I think we should move to merge this PR since we'll want it for checkpoints to work.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **d88943e**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #634 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/d6cG3rERk89DQQxYoI9InMs0PqZrBUKo)

**Comment by no2chem** [comment]
Ok, perf looks good, will merge after Travis passes.
